bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryCredentialsSupplier.java

```diff
@@ -53,25 +70,97 @@ public BigQueryCredentialsSupplier(
-          new AccessTokenProviderCredentials(verifySerialization(accessTokenProvider));
+      credentials = new AccessTokenProviderCredentials(verifySerialization(accessTokenProvider));
     } else if (accessToken.isPresent()) {
-      this.credentials = createCredentialsFromAccessToken(accessToken.get());
+      credentials = createCredentialsFromAccessToken(accessToken.get());
     } else if (credentialsKey.isPresent()) {
-      this.credentials =
+      credentials =
           createCredentialsFromKey(credentialsKey.get(), proxyUri, proxyUsername, proxyPassword);
     } else if (credentialsFile.isPresent()) {
-      this.credentials =
+      credentials =
           createCredentialsFromFile(credentialsFile.get(), proxyUri, proxyUsername, proxyPassword);
     } else {
-      this.credentials = createDefaultCredentials();
+      credentials = createDefaultCredentials();
     }
+    Credentials impersonatedCredentials =
+        createCredentialsFromImpersonation(
+            loggedInUserName,
+            loggedInUserGroups,
+            impersonationServiceAccountsForUsers,
+            impersonationServiceAccountsForGroups,
+            impersonationServiceAccount,
+            (GoogleCredentials) credentials,
+            proxyUri,
+            proxyUsername,
+            proxyPassword);
+    if (impersonatedCredentials != null) {
+      credentials = impersonatedCredentials;
+    }
+    this.credentials = credentials;
   }
 
   private static Credentials createCredentialsFromAccessToken(String accessToken) {
     return GoogleCredentials.create(new AccessToken(accessToken, null));
   }
 
+  private static Credentials createCredentialsFromImpersonation(
+      Optional<String> loggedInUserName,
+      Optional<Set<String>> loggedInUserGroups,
+      Optional<Map<String, String>> impersonationServiceAccountsForUsers,
+      Optional<Map<String, String>> impersonationServiceAccountsForGroups,
+      Optional<String> impersonationServiceAccount,
+      GoogleCredentials sourceCredentials,
+      Optional<URI> proxyUri,
+      Optional<String> proxyUsername,
+      Optional<String> proxyPassword) {
+
+    Optional<String> serviceAccountToImpersonate =
+        Stream.of(
+                () ->
+                    getServiceAccountToImpersonateByKeys(
+                        impersonationServiceAccountsForUsers,
+                        loggedInUserName.map(ImmutableList::of).orElseGet(ImmutableList::of)),
+                () ->
+                    getServiceAccountToImpersonateByKeys(
+                        impersonationServiceAccountsForGroups,
+                        loggedInUserGroups.orElse(new HashSet<>())),
+                (Supplier<Optional<String>>) () -> impersonationServiceAccount)
+            .map(Supplier::get)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .filter(sa -> !isNullOrEmpty(sa))
+            .findFirst();
+
+    if (serviceAccountToImpersonate.isPresent()) {
```

Comment:
Should be `return serviceAccountToImpersonate.map(sa -> ImpersonatedCredentials.newBuilder().....build());`